src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs

```diff
@@ -477,6 +477,11 @@ private static bool HaveSameReturnTypes(Symbol member1, TypeMap typeMap1, Symbol
                 return false;
             }
 
+            if (unsubstitutedReturnType1.Type is null || unsubstitutedReturnType2.Type is null)
```

Comment:
@RikkiGibson @agocke are probably the best to review this, but I'm unclear as to why the unsubstitutedReturnTypes here would null. That doesn't seem like it should be a case we can get into with valid code.

Reply:
Or more appropriately, _should_ get into with valid code.

Reply:
Yeah, it's very difficult to see how an overridden method could have a `null` ReturnType. I think that's the root of the issue we need to look at.

Reply:
I don't think you're constructing a malformed symbol table. I think you're just unexpectedly attempting to use it at a point in time when it's not yet ready.

The disparity starts at this line in `Microsoft.CodeAnalysis.CSharp.Symbols.SourceOrdinaryMethodSymbol.MethodChecks`:

```cs
_lazyReturnType = signatureBinder.BindType(returnTypeSyntax, diagnostics);
```

In other cases, `BindType` returns successfully with a result. In the where the return type is `T*`, the call to `BindType` eventually goes through `Microsoft.CodeAnalysis.CSharp.Binder.BindNamespaceOrTypeOrAliasSymbol.__bindPointer`. That function uses the `Microsoft.CodeAnalysis.CSharp.Symbols.TypeParameterSymbol.ManagedKind` getter. Now that getter uses the `Microsoft.CodeAnalysis.CSharp.Symbols.TypeParameterSymbol.HasUnmanagedTypeConstraint` getter which (in the case of an override) tries to consult the overridden method.

To find which method is being overridden, you use `Microsoft.CodeAnalysis.CSharp.Symbols.OverriddenOrHiddenMembersHelpers.FindOverriddenOrHiddenMembersInType`, which tries to match the base class methods by various criteria, including the return type. But the return type (`Microsoft.CodeAnalysis.CSharp.Symbols.SourceOrdinaryMethodSymbol._lazyReturnType`) has not been assigned yet.

I personally see two ways of solving this:

- Somehow make `BindType` (specifically `__bindPointer()`) work without depending on the return type when resolving an override. This might be tricky, because the part of this code path that deals with the generic pointer does not know/care that it's resolving an override, and the part deals with the override does not know/care that it's resolving a generic pointer.
- Accept this as a possible code path and make the `exactMatchComparer` more "forgiving" in the presence of missing information. Not forgiving in the sense of returning a false positive, of course, but forgiving in the sense of returning instead of throwing.

The latter solution does indeed feel like a cop out -- but at least it shouldn't be breaking previously working code, because it only deals with a case that would have thrown an exception before. Still, I will spend some time over the weekend trying to find out if I can rewire things to implement something closer to the first solution.

Reply:
IIRC the reason we check ManagedKind when binding a pointer is in order to check the constraint that the pointed-at type must be unmanaged. So maybe we should delay this check until after the method has been fully bound--i.e. ReturnType and other important properties are populated.

Reply:
Yeah, this is definitely wrong, if we have a null type in hand at the time we're doing this check it means we either haven't frontloaded the binding enough, or we're doing the check too early.